File: src/api/schemas/stories.py
"""Pydantic schemas for Storyline & Graph API responses."""
from typing import Optional
from pydantic import BaseModel


class StorylineNode(BaseModel):
    """A storyline node in the narrative graph."""
    id: int
    title: str
    summary: Optional[str] = None
    category: Optional[str] = None
    narrative_status: str  # emerging, active, stabilized
    momentum_score: float
    article_count: int
    key_entities: list[str] = []
    start_date: Optional[str] = None
    last_update: Optional[str] = None
    days_active: Optional[int] = None
    community_id: Optional[int] = None


class StorylineEdge(BaseModel):
    """An edge between two storylines."""
    source: int
    target: int
    weight: float
    relation_type: str = "relates_to"


class GraphStats(BaseModel):
    """Aggregate stats for the graph."""
    total_nodes: int
    total_edges: int
    avg_momentum: float
    communities_count: int = 0
    avg_edges_per_node: float = 0.0


class CommunityInfo(BaseModel):
    """Summary of a Louvain graph community."""
    community_id: int
    size: int
    label: str
    top_storylines: list["StorylineNode"] = []
    key_entities: list[str] = []
    avg_momentum: float = 0.0


class GraphNetwork(BaseModel):
    """Full graph network: nodes + links + stats."""
    nodes: list[StorylineNode]
    links: list[StorylineEdge]
    stats: GraphStats


class RelatedStoryline(BaseModel):
    """A related storyline reference."""
    id: int
    title: str
    weight: float
    relation_type: str


class LinkedArticle(BaseModel):
    """An article linked to a storyline."""
    id: int
    title: str
    source: Optional[str] = None
    published_date: Optional[str] = None


class StorylineDetail(BaseModel):
    """Detailed storyline with relations and articles."""
    storyline: StorylineNode
    related_storylines: list[RelatedStoryline] = []
    recent_articles: list[LinkedArticle] = []

File: web-platform/types/stories.ts
// Storyline & Graph API Response Types

export type NarrativeStatus = 'emerging' | 'active' | 'stabilized';

export interface StorylineNode {
  id: number;
  title: string;
  summary: string | null;
  category: string | null;
  narrative_status: NarrativeStatus;
  momentum_score: number;
  article_count: number;
  key_entities: string[];
  start_date: string | null;
  last_update: string | null;
  days_active: number | null;
  community_id?: number | null;
  community_name?: string | null;
}

export interface StorylineEdge {
  source: number;
  target: number;
  weight: number;
  relation_type: string;
}

export interface GraphStats {
  total_nodes: number;
  total_edges: number;
  avg_momentum: number;
  communities_count: number;
  avg_edges_per_node: number;
}

export interface GraphNetwork {
  nodes: StorylineNode[];
  links: StorylineEdge[];
  stats: GraphStats;
}

export interface GraphNetworkResponse {
  success: boolean;
  data: GraphNetwork;
  error: string | null;
  generated_at: string;
}

// Detail types

export interface RelatedStoryline {
  id: number;
  title: string;
  weight: number;
  relation_type: string;
}

export interface LinkedArticle {
  id: number;
  title: string;
  source: string | null;
  published_date: string | null;
  bullet_points?: string[];
}

export interface StorylineDetailData {
  storyline: StorylineNode;
  related_storylines: RelatedStoryline[];
  recent_articles: LinkedArticle[];
}

export interface StorylineDetailResponse {
  success: boolean;
  data: StorylineDetailData;
  error: string | null;
  generated_at: string;
}

// Ego network (returned by /stories/{id}/network)
export interface EgoNetworkData {
  center_node: StorylineNode;
  neighbors: StorylineNode[];
  edges: StorylineEdge[];
}

export interface EgoNetworkResponse {
  success: boolean;
  data: EgoNetworkData;
  generated_at: string;
}
